clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -290,17 +315,47 @@ public void poll(long timeout) throws IOException {
 
         clear();
 
-        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty())
+        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty() || !keysWithBytesBuffered.isEmpty())
             timeout = 0;
 
+        boolean poolOutOfMemory = memoryPool.isOutOfMemory();
+        if (poolOutOfMemory && !outOfMemory) {
+            //dont bother selecting for reads
+            outOfMemory = true;
+            log.trace("Broker is low on memory - muting incoming sockets");
+            for (KafkaChannel channel : channels.values()) {
+                if (channel.isInMutableState())
+                    channel.mute();
+            }
```

Comment:
@ewencp The code looks like it is proactively closing most channels. But actually it closes a small subset of channels. Channels can be in one of these states:
1. Handshake
2. Authentication
3. Waiting to receive a message (receive == null)
4. Received partial message size (receive != null, buffer == null)
5. Received size and partial message body (receive != null, buffer != null)
6. Muted after receiving size due to OOM
7. Explicitly muted
8. Disconnect

The loop actually handles only 4). It mutes 2) at the moment, but that is pointless since authentication doesn't use the pool, so that needs fixing anyway. 4) already has the size buffer, so there is not much point in muting before size is read, after which it will move to 6) if still OOM. Muting proactively is not particularly helpful since disconnect processing gets delayed as well, hence 3) is not muted. If we decide to allocate small buffers outside the pool to handle consumers as Mickael has suggested, it will be useful to mute only in one place - i.e. when a buffer needs to get allocated and its size is known. I think `isInMutableState`
is unnecessary if muting is done on allocation failure and that makes the code simpler.